cirq-core/cirq/transformers/measurement_transformers.py

```diff
@@ -227,3 +229,52 @@ def flip_inversion(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
     return transformer_primitives.map_operations(
         circuit, flip_inversion, deep=context.deep if context else True, tags_to_ignore=ignored
     ).unfreeze()
+
+
+class _CX(ops.Gate):
+    """A CX gate generalized for qudits.
+
+    This represents a Controlled-NOT gate for qudits, in the following sense. If the target is
+    zero, then it becomes the value of the control. If the target is equal to the control, then
+    it becomes zero. All other cases do not affect the target. The first rule is the only one
+    required for measurement-deferral purposes: the ancilla qudit representing the creg is always
+    zero at the time of measurement. The remaining rules are for symmetry.
+
+    |k0> -> |kk>
+    |kk> -> |k0>
+    |kj> -> |kj> otherwise
```

Comment:
This is a weird gate. It would be better to use a well-known operation such as modular addition that acts as follows on two qudits
```
|a,b> -> |a,a+b mod d>
```
where d is the dimension of the Hilbert space of each qudit and a,b=0,1,...d-1. Note that CNOT is a special case of the above gate for d=2.

Modular addition satisfies the first rule. It also satisfies the second rule when d is even. It satisfies the third only in the case when k=0.

Reply:
Ah, bummer I liked the `_CX` and its `CX∘CX == I, and CX∘XC∘CX == SWAP` properties.

But I think the adder gate is better: even classically, if the creg is _not_ zero (say we add some feature allowing measurement to XOR the creg), then in the multi-dimensional extension, that `_CX` logic is really weird classically. Whereas adding makes good sense.